src/Orleans/Streams/Internal/StreamImpl.cs

```diff
@@ -132,9 +137,19 @@ public Task<IList<StreamSubscriptionHandle<T>>> GetAllSubscriptionHandles()
 
         internal Task UnsubscribeAsync(StreamSubscriptionHandle<T> handle)
         {
+            CheckExecutingContext("UnsubscribeAsync");
             return GetConsumerInterface().UnsubscribeAsync(handle);
         }
 
+        private void CheckExecutingContext(string caller)
+        {
+            var context = GetStreamProviderRuntime().GetCurrentSchedulingContext();
+            if (context != null && !context.ContextType.Equals(SchedulingContextType.Activation))
+            {
+                throw new OrleansException(String.Format("Attemping to {0} on a stream not from within Grain context.", caller));
```

Comment:
What about from Orleans clients?
Attemping == Attempting
Also, clumsy wording.  Maybe "Attempting to {0} on a stream from outside a Grain context."?